File: nestack-backend/src/modules/finance/finance.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Between, LessThanOrEqual, MoreThanOrEqual } from 'typeorm';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';
import {
  BankAccount,
  Transaction,
  OpenBankingToken,
  User,
} from '../../database/entities';
import { TransactionType, ShareStatus } from '../../common/enums';
import {
  BankAccountNotFoundException,
  OpenBankingTokenNotFoundException,
  UnauthorizedAccessException,
} from '../../common/exceptions/business.exception';
import {
  AccountResponseDto,
  UpdateAccountDto,
  TransactionResponseDto,
  TransactionFilterDto,
  OpenBankingStatusDto,
} from './dto';
import { PaginatedResponse, PaginationMeta } from '../../common/dto/api-response.dto';

@Injectable()
export class FinanceService {
  private readonly logger = new Logger(FinanceService.name);

  constructor(
    @InjectRepository(BankAccount)
    private bankAccountRepository: Repository<BankAccount>,
    @InjectRepository(Transaction)
    private transactionRepository: Repository<Transaction>,
    @InjectRepository(OpenBankingToken)
    private openBankingTokenRepository: Repository<OpenBankingToken>,
    @InjectRepository(User)
    private userRepository: Repository<User>,
    private httpService: HttpService,
    private configService: ConfigService,
  ) {}

  // Open Banking
  getAuthorizeUrl(userId: string): string {
    const baseUrl = this.configService.get('openbanking.baseUrl');
    const clientId = this.configService.get('openbanking.clientId');
    const redirectUri = this.configService.get('openbanking.redirectUri');

    const state = Buffer.from(JSON.stringify({ userId })).toString('base64');

    return `${baseUrl}/oauth/2.0/authorize?response_type=code&client_id=${clientId}&redirect_uri=${encodeURIComponent(redirectUri)}&scope=login inquiry transfer&state=${state}&auth_type=0`;
  }

  async handleCallback(code: string, state: string): Promise<void> {
    const { userId } = JSON.parse(Buffer.from(state, 'base64').toString());

    const baseUrl = this.configService.get('openbanking.baseUrl');
    const clientId = this.configService.get('openbanking.clientId');
    const clientSecret = this.configService.get('openbanking.clientSecret');
    const redirectUri = this.configService.get('openbanking.redirectUri');

    try {
      // Exchange code for token
      const tokenResponse = await firstValueFrom(
        this.httpService.post(`${baseUrl}/oauth/2.0/token`, null, {
          params: {
            code,
            client_id: clientId,
            client_secret: clientSecret,
            redirect_uri: redirectUri,
            grant_type: 'authorization_code',
          },
        }),
      );

      const tokenData = tokenResponse.data;

      // Save or update token
      let token = await this.openBankingTokenRepository.findOne({
        where: { userId },
      });

      if (token) {
        token.accessToken = tokenData.access_token;
        token.refreshToken = tokenData.refresh_token;
        token.tokenType = tokenData.token_type;
        token.scope = tokenData.scope;
        token.userSeqNo = tokenData.user_seq_no;
        token.expiresAt = new Date(Date.now() + tokenData.expires_in * 1000);
      } else {
        token = this.openBankingTokenRepository.create({
          userId,
          accessToken: tokenData.access_token,
          refreshToken: tokenData.refresh_token,
          tokenType: tokenData.token_type,
          scope: tokenData.scope,
          userSeqNo: tokenData.user_seq_no,
          expiresAt: new Date(Date.now() + tokenData.expires_in * 1000),
        });
      }

      await this.openBankingTokenRepository.save(token);

      // Sync accounts
      await this.syncAccounts(userId);
    } catch (error) {
      this.logger.error('Open Banking callback failed', error);
      throw error;
    }
  }

  async getOpenBankingStatus(userId: string): Promise<OpenBankingStatusDto> {
    const token = await this.openBankingTokenRepository.findOne({
      where: { userId },
    });

    if (!token) {
      return { isConnected: false };
    }

    return {
      isConnected: true,
      connectedAt: token.createdAt,
      expiresAt: token.expiresAt,
    };
  }

  async disconnectOpenBanking(userId: string): Promise<void> {
    await this.openBankingTokenRepository.delete({ userId });
    // Optionally delete accounts and transactions
    // await this.bankAccountRepository.delete({ userId });
  }

  // Accounts
  async getAccounts(userId: string): Promise<AccountResponseDto[]> {
    const accounts = await this.bankAccountRepository.find({
      where: { userId, isHidden: false },
      order: { createdAt: 'DESC' },
    });

    return accounts.map((a) => AccountResponseDto.fromEntity(a));
  }

  async syncAccounts(userId: string): Promise<AccountResponseDto[]> {
    const token = await this.openBankingTokenRepository.findOne({
      where: { userId },
    });

    if (!token) {
      throw new OpenBankingTokenNotFoundException();
    }

    // In production, call Open Banking API to get accounts
    // For now, return existing accounts
    this.logger.log(`Syncing accounts for user ${userId}`);

    const accounts = await this.bankAccountRepository.find({
      where: { userId },
    });

    // Update last synced time
    await this.bankAccountRepository.update(
      { userId },
      { lastSyncedAt: new Date() },
    );

    return accounts.map((a) => AccountResponseDto.fromEntity(a));
  }

  async updateAccount(
    userId: string,
    accountId: string,
    dto: UpdateAccountDto,
  ): Promise<AccountResponseDto> {
    const account = await this.bankAccountRepository.findOne({
      where: { id: accountId, userId },
    });

    if (!account) {
      throw new BankAccountNotFoundException();
    }

    if (dto.accountAlias !== undefined) account.accountAlias = dto.accountAlias;
    if (dto.shareStatus !== undefined) account.shareStatus = dto.shareStatus;
    if (dto.isHidden !== undefined) account.isHidden = dto.isHidden;

    await this.bankAccountRepository.save(account);

    return AccountResponseDto.fromEntity(account);
  }

  // Transactions
  async getTransactions(
    userId: string,
    filters: TransactionFilterDto,
  ): Promise<PaginatedResponse<TransactionResponseDto>> {
    const page = filters.page || 1;
    const limit = filters.limit || 20;
    const skip = (page - 1) * limit;

    // Get user's account IDs
    const accounts = await this.bankAccountRepository.find({
      where: { userId },
      select: ['id'],
    });

    const accountIds = accounts.map((a) => a.id);

    if (accountIds.length === 0) {
      return new PaginatedResponse([], new PaginationMeta(page, limit, 0));
    }

    const queryBuilder = this.transactionRepository
      .createQueryBuilder('transaction')
      .where('transaction.bank_account_id IN (:...accountIds)', { accountIds });

    if (filters.accountId) {
      queryBuilder.andWhere('transaction.bank_account_id = :accountId', {
        accountId: filters.accountId,
      });
    }

    if (filters.type) {
      queryBuilder.andWhere('transaction.type = :type', { type: filters.type });
    }

    if (filters.startDate && filters.endDate) {
      queryBuilder.andWhere(
        'transaction.transaction_date BETWEEN :startDate AND :endDate',
        { startDate: filters.startDate, endDate: filters.endDate },
      );
    } else if (filters.startDate) {
      queryBuilder.andWhere('transaction.transaction_date >= :startDate', {
        startDate: filters.startDate,
      });
    } else if (filters.endDate) {
      queryBuilder.andWhere('transaction.transaction_date <= :endDate', {
        endDate: filters.endDate,
      });
    }

    if (filters.missionId) {
      queryBuilder.andWhere('transaction.mission_id = :missionId', {
        missionId: filters.missionId,
      });
    }

    queryBuilder.orderBy('transaction.transaction_date', 'DESC');
    queryBuilder.addOrderBy('transaction.transaction_time', 'DESC');
    queryBuilder.skip(skip).take(limit);

    const [transactions, totalItems] = await queryBuilder.getManyAndCount();

    const data = transactions.map((t) => TransactionResponseDto.fromEntity(t));
    const meta = new PaginationMeta(page, limit, totalItems);

    return new PaginatedResponse(data, meta);
  }

  async syncTransactions(userId: string, accountId: string): Promise<void> {
    const account = await this.bankAccountRepository.findOne({
      where: { id: accountId, userId },
    });

    if (!account) {
      throw new BankAccountNotFoundException();
    }

    const token = await this.openBankingTokenRepository.findOne({
      where: { userId },
    });

    if (!token) {
      throw new OpenBankingTokenNotFoundException();
    }

    // In production, call Open Banking API to get transactions
    this.logger.log(`Syncing transactions for account ${accountId}`);

    // Update last synced time
    account.lastSyncedAt = new Date();
    await this.bankAccountRepository.save(account);
  }
}

File: nestack-backend/src/main/kotlin/com/nestack/domain/finance/service/FinanceService.kt
package com.nestack.domain.finance.service

import com.nestack.common.constant.ErrorCode
import com.nestack.common.enum.TransactionType
import com.nestack.common.exception.BusinessException
import com.nestack.common.util.CryptoUtil
import com.nestack.domain.finance.dto.TransactionQueryRequest
import com.nestack.domain.finance.dto.UpdateAccountRequest
import com.nestack.infrastructure.persistence.entity.BankAccount
import com.nestack.infrastructure.persistence.entity.OpenBankingToken
import com.nestack.infrastructure.persistence.entity.Transaction
import com.nestack.infrastructure.persistence.repository.BankAccountRepository
import com.nestack.infrastructure.persistence.repository.OpenBankingTokenRepository
import com.nestack.infrastructure.persistence.repository.TransactionRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*

@Service
class FinanceService(
    private val bankAccountRepository: BankAccountRepository,
    private val transactionRepository: TransactionRepository,
    private val tokenRepository: OpenBankingTokenRepository
) {
    private val logger = LoggerFactory.getLogger(FinanceService::class.java)

    @Value("\${spring.openbanking.api-url:https://testapi.openbanking.or.kr}")
    private var apiBaseUrl: String = "https://testapi.openbanking.or.kr"

    @Value("\${spring.openbanking.client-id:}")
    private var clientId: String = ""

    @Value("\${spring.openbanking.client-secret:}")
    private var clientSecret: String = ""

    @Value("\${spring.openbanking.redirect-uri:}")
    private var redirectUri: String = ""

    @Value("\${spring.encryption.key:}")
    private var encryptionKey: String = ""

    fun getAuthorizationUrl(state: String): String {
        val params = mapOf(
            "response_type" to "code",
            "client_id" to clientId,
            "redirect_uri" to redirectUri,
            "scope" to "login inquiry transfer",
            "state" to state,
            "auth_type" to "0"
        )

        val queryString = params.entries.joinToString("&") { "${it.key}=${it.value}" }
        return "$apiBaseUrl/oauth/2.0/authorize?$queryString"
    }

    @Transactional
    fun handleCallback(userId: UUID, code: String) {
        try {
            // Exchange authorization code for tokens (simulated)
            val tokenResponse = exchangeCodeForTokens(code)

            // Encrypt tokens
            val encryptedAccessToken = CryptoUtil.encrypt(tokenResponse.accessToken, encryptionKey)
            val encryptedRefreshToken = CryptoUtil.encrypt(tokenResponse.refreshToken, encryptionKey)

            // Store or update tokens
            val existingToken = tokenRepository.findByUserId(userId)

            if (existingToken != null) {
                existingToken.accessToken = encryptedAccessToken
                existingToken.refreshToken = encryptedRefreshToken
                existingToken.tokenType = tokenResponse.tokenType
                existingToken.scope = tokenResponse.scope
                existingToken.userSeqNo = tokenResponse.userSeqNo
                existingToken.expiresAt = LocalDateTime.now().plusSeconds(tokenResponse.expiresIn)
                tokenRepository.save(existingToken)
            } else {
                val newToken = OpenBankingToken().apply {
                    this.userId = userId
                    this.accessToken = encryptedAccessToken
                    this.refreshToken = encryptedRefreshToken
                    this.tokenType = tokenResponse.tokenType
                    this.scope = tokenResponse.scope
                    this.userSeqNo = tokenResponse.userSeqNo
                    this.expiresAt = LocalDateTime.now().plusSeconds(tokenResponse.expiresIn)
                }
                tokenRepository.save(newToken)
            }

            // Sync accounts
            syncAccounts(userId)
        } catch (e: Exception) {
            logger.error("OpenBanking callback error", e)
            throw BusinessException(ErrorCode.FINANCE_001)
        }
    }

    private fun exchangeCodeForTokens(code: String): TokenResponse {
        // In production, make actual HTTP call to OpenBanking API
        logger.info("Exchanging code for tokens: $code")

        // Simulated response
        return TokenResponse(
            accessToken = "access_${System.currentTimeMillis()}",
            refreshToken = "refresh_${System.currentTimeMillis()}",
            tokenType = "Bearer",
            expiresIn = 7776000L, // 90 days
            scope = "login inquiry transfer",
            userSeqNo = "U${System.currentTimeMillis()}"
        )
    }

    @Transactional
    fun syncAccounts(userId: UUID): List<BankAccount> {
        val token = getValidToken(userId)
            ?: throw BusinessException(ErrorCode.FINANCE_002)

        try {
            // Fetch accounts from OpenBanking API (simulated)
            val accountsData = fetchAccountsFromOpenBanking(token)

            val accounts = mutableListOf<BankAccount>()

            accountsData.forEach { accountData ->
                var account = bankAccountRepository.findByUserId(userId)
                    .find { it.fintechUseNum == accountData.fintechUseNum }

                if (account == null) {
                    account = BankAccount().apply {
                        this.userId = userId
                        this.bankCode = accountData.bankCode
                        this.bankName = accountData.bankName
                        this.accountNumber = CryptoUtil.encrypt(accountData.accountNumber, encryptionKey)
                        this.accountNumberMasked = accountData.accountNumberMasked
                        this.accountType = accountData.accountType
                        this.balance = accountData.balance
                        this.fintechUseNum = accountData.fintechUseNum
                        this.lastSyncedAt = LocalDateTime.now()
                    }
                } else {
                    account.balance = accountData.balance
                    account.lastSyncedAt = LocalDateTime.now()
                }

                accounts.add(bankAccountRepository.save(account))
            }

            return accounts
        } catch (e: Exception) {
            logger.error("Error syncing accounts", e)
            throw BusinessException(ErrorCode.FINANCE_001)
        }
    }

    @Transactional(readOnly = true)
    fun isConnected(userId: UUID): Boolean {
        val token = getValidToken(userId)
        return token != null
    }

    @Transactional
    fun disconnect(userId: UUID) {
        tokenRepository.findByUserId(userId)?.let {
            tokenRepository.delete(it)
        }
    }

    @Transactional(readOnly = true)
    fun getAccounts(userId: UUID): List<BankAccount> {
        return bankAccountRepository.findByUserId(userId)
            .filter { !it.isHidden }
            .sortedByDescending { it.createdAt }
    }

    @Transactional(readOnly = true)
    fun getAccountById(userId: UUID, accountId: UUID): BankAccount {
        return bankAccountRepository.findByUserIdAndId(userId, accountId)
            ?: throw BusinessException(ErrorCode.FINANCE_003)
    }

    @Transactional
    fun updateAccount(userId: UUID, accountId: UUID, request: UpdateAccountRequest): BankAccount {
        val account = getAccountById(userId, accountId)

        request.accountAlias?.let { account.accountAlias = it }
        request.shareStatus?.let { account.shareStatus = it }
        request.isHidden?.let { account.isHidden = it }

        return bankAccountRepository.save(account)
    }

    @Transactional
    fun syncAccountTransactions(userId: UUID, accountId: UUID): List<Transaction> {
        val account = getAccountById(userId, accountId)
        val token = getValidToken(userId)
            ?: throw BusinessException(ErrorCode.FINANCE_002)

        try {
            // Fetch transactions from OpenBanking API (simulated)
            val transactionsData = fetchTransactionsFromOpenBanking(token, account.fintechUseNum)

            val transactions = mutableListOf<Transaction>()

            transactionsData.forEach { txData ->
                val existingTx = transactionRepository.findByBankAccountId(account.id)
                    .find { it.transactionId == txData.transactionId }

                if (existingTx == null) {
                    val tx = Transaction().apply {
                        this.bankAccountId = account.id
                        this.transactionId = txData.transactionId
                        this.type = if (txData.inoutType == "I") TransactionType.DEPOSIT else TransactionType.WITHDRAWAL
                        this.amount = txData.amount
                        this.balanceAfter = txData.balanceAfter
                        this.description = txData.description
                        this.counterparty = txData.counterparty
                        this.transactionDate = txData.transactionDate
                        this.transactionTime = txData.transactionTime
                    }
                    transactions.add(transactionRepository.save(tx))
                }
            }

            // Update account balance
            if (transactionsData.isNotEmpty()) {
                account.balance = transactionsData.first().balanceAfter
                account.lastSyncedAt = LocalDateTime.now()
                bankAccountRepository.save(account)
            }

            return transactions
        } catch (e: Exception) {
            logger.error("Error syncing transactions", e)
            throw BusinessException(ErrorCode.FINANCE_001)
        }
    }

    @Transactional(readOnly = true)
    fun getTransactions(userId: UUID, accountId: UUID?, query: TransactionQueryRequest): List<Transaction> {
        val accountIds = if (accountId != null) {
            listOf(accountId)
        } else {
            bankAccountRepository.findByUserId(userId).map { it.id }
        }

        return if (query.startDate != null && query.endDate != null) {
            transactionRepository.findByBankAccountIdsAndDateRange(accountIds, query.startDate, query.endDate)
        } else {
            transactionRepository.findByBankAccountIds(accountIds)
        }.filter { query.type == null || it.type == query.type }
    }

    private fun getValidToken(userId: UUID): OpenBankingToken? {
        val token = tokenRepository.findByUserId(userId) ?: return null
        if (token.expiresAt.isBefore(LocalDateTime.now())) {
            return null
        }
        return token
    }

    private fun fetchAccountsFromOpenBanking(token: OpenBankingToken): List<AccountData> {
        // Simulated - in production, make actual API call
        return listOf(
            AccountData(
                fintechUseNum = "F${System.currentTimeMillis()}_1",
                accountNumber = "1234567890123456",
                accountNumberMasked = "1234-****-****-3456",
                bankCode = "004",
                bankName = "KB국민은행",
                accountType = "입출금",
                balance = java.math.BigDecimal(1500000)
            )
        )
    }

    private fun fetchTransactionsFromOpenBanking(token: OpenBankingToken, fintechUseNum: String): List<TransactionData> {
        // Simulated - in production, make actual API call
        return emptyList()
    }

    private data class TokenResponse(
        val accessToken: String,
        val refreshToken: String,
        val tokenType: String,
        val expiresIn: Long,
        val scope: String,
        val userSeqNo: String
    )

    private data class AccountData(
        val fintechUseNum: String,
        val accountNumber: String,
        val accountNumberMasked: String,
        val bankCode: String,
        val bankName: String,
        val accountType: String,
        val balance: java.math.BigDecimal
    )

    private data class TransactionData(
        val transactionId: String,
        val inoutType: String,
        val amount: java.math.BigDecimal,
        val balanceAfter: java.math.BigDecimal,
        val description: String?,
        val counterparty: String?,
        val transactionDate: java.time.LocalDate,
        val transactionTime: java.time.LocalTime?
    )
}
